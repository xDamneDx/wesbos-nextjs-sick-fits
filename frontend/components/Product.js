import Link from 'next/link';
import formatMoney from '../lib/formatMoney';

// Components:
import DeleteProduct from './DeleteProduct';

// Styled components:
import ItemStyles from './styles/ItemStyles';
import TitleStyles from './styles/TitleStyles';
import PriceTagStyles from './styles/PriceTagStyles';

function Product({ product }) {
  return (
    <ItemStyles>
      <img src={product.photo.image.publicUrlTransformed} alt={product.name} />
      <TitleStyles>
        <Link href={`/product/${product.id}`}>{product.name}</Link>
      </TitleStyles>
      <PriceTagStyles>{formatMoney(product.price)}</PriceTagStyles>
      <p>{product.description}</p>
      <div className="buttonList">
        <Link
          href={{
            pathname: '/update',
            query: {
              id: product.id,
            },
          }}
        >
          Edit ✏️
        </Link>
        <DeleteProduct id={product.id}>Delete</DeleteProduct>
      </div>
    </ItemStyles>
  );
}

export default Product;
