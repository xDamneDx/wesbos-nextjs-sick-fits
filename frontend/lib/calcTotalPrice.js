export default function calcTotalPrice(cart) {
  return cart.reduce((tally, cartItem) => {
    // Products can be deleted, but they could still be in your cart
    if (!cartItem.product) return tally;

    return tally + cartItem.quantity * cartItem.product.price;
  }, 0);
}
