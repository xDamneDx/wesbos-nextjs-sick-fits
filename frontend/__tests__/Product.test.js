import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import { fakeItem } from '../lib/testUtils';
import Product from '../components/Product';

const product = fakeItem();

describe('<Product/>', () => {
  it('renders out the price tag and title', () => {
    const { container, debug } = render(
      <MockedProvider>
        <Product product={product} />
      </MockedProvider>
    );

    const priceTag = screen.getByText('$50');
    expect(priceTag).toBeInTheDocument();

    const link = container.querySelector('a');
    expect(link).toHaveAttribute('href', '/product/abc123');
    expect(link).toHaveTextContent(product.name);
  });

  it('Renders anf matches the snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <Product product={product} />
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
