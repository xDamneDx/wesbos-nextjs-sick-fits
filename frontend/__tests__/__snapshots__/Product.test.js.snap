// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product/> Renders anf matches the snapshot 1`] = `
<div>
  <div
    class="ItemStyles-sc-16pk14u-0 fAcGEr"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <h3
      class="TitleStyles-sc-ivt2zh-0 cpTxwY"
    >
      <a
        href="/product/abc123"
      >
        dogs are best
      </a>
    </h3>
    <span
      class="PriceTagStyles-sc-1rye5zs-0 fiAbGV"
    >
      $50
    </span>
    <p>
      dogs
    </p>
    <div
      class="buttonList"
    >
      <a
        href="/update?id=abc123"
      >
        Edit ✏️
      </a>
      <button
        type="button"
      >
        Add
         To Cart 🛒
      </button>
      <button
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
